File: ios/Runner/Plugin/AzureLivenessPlugin.swift
import Flutter
import UIKit

/// Platform channel bridge between Flutter and the native Azure AI Face Liveness SDK.
///
/// The Azure Face Liveness SDK for iOS requires:
///   1. An access token obtained from a backend session creation endpoint.
///   2. A ViewController-based UI that guides the user through head movement
///      and lighting checks for Presentation Attack Detection (PAD).
///   3. A verification payload returned asynchronously via delegate callback.
///
/// Implementation steps (requires AzureAIFaceSDK pod):
///   1. Add `pod 'AzureAIFaceSDK'` to the Podfile.
///   2. Create a `LivenessWithVerifySessionViewController` with the access token.
///   3. Present the view controller and handle the delegate callback to
///      obtain the verifySessionId and the encrypted verification payload.
class AzureLivenessPlugin: NSObject, FlutterPlugin {

    static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(
            name: "com.dogsafield.dogsafield/azure_liveness",
            binaryMessenger: registrar.messenger()
        )
        let instance = AzureLivenessPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
    }

    func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "startLivenessSession":
            guard let args = call.arguments as? [String: Any],
                  let authToken = args["authToken"] as? String else {
                result(FlutterError(
                    code: "INVALID_ARGS",
                    message: "authToken is required",
                    details: nil
                ))
                return
            }
            startLivenessSession(authToken: authToken, result: result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private func startLivenessSession(authToken: String, result: @escaping FlutterResult) {
        // Native Azure Face Liveness SDK integration:
        //
        // 1. Initialize the SDK:
        //    let sdk = FaceLivenessSessionController(
        //        authToken: authToken,
        //        delegate: self
        //    )
        //
        // 2. Present the liveness camera UI from the root view controller:
        //    guard let controller = UIApplication.shared
        //        .keyWindow?.rootViewController else { return }
        //    controller.present(sdk, animated: true)
        //
        // 3. Handle the delegate callback:
        //    func livenessSessionController(
        //        _ controller: FaceLivenessSessionController,
        //        didCompleteWithResult result: LivenessWithVerifySessionResult
        //    ) {
        //        let verifySessionId = result.sessionId
        //        let payload = result.encryptedVerificationPayload
        //        // forward back to Flutter via result(verifySessionId)
        //    }
        //
        // This stub returns the authToken as a simulated verifySessionId.
        result(authToken)
    }
}

File: android/app/src/main/kotlin/com/dogsafield/dogsafield/plugin/AzureLivenessPlugin.kt
package com.dogsafield.dogsafield.plugin

import android.app.Activity
import io.flutter.embedding.engine.FlutterEngine
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

/**
 * Platform channel bridge between Flutter and the native Azure AI Face Liveness SDK.
 *
 * The Azure Face Liveness SDK requires:
 *   1. An access token obtained from a backend session creation endpoint.
 *   2. A native Activity-based UI that guides the user through head movement
 *      and lighting checks for Presentation Attack Detection (PAD).
 *   3. A verification payload returned asynchronously.
 *
 * Implementation steps (requires com.azure:azure-ai-face-android-rest-internal):
 *   1. Add the Azure Face SDK dependency to app/build.gradle.kts.
 *   2. Call FaceSessionClient.createLivenessWithVerifySession() with the access token.
 *   3. Launch LivenessWithVerifySessionIntents.createSessionIntent() to start
 *      the native camera overlay.
 *   4. Handle the onActivityResult callback to extract the verifySessionId and
 *      the encrypted verification payload.
 */
class AzureLivenessPlugin(private val activity: Activity) :
    MethodChannel.MethodCallHandler {

    companion object {
        private const val CHANNEL = "com.dogsafield.dogsafield/azure_liveness"

        fun registerWith(flutterEngine: FlutterEngine, activity: Activity) {
            val channel = MethodChannel(
                flutterEngine.dartExecutor.binaryMessenger,
                CHANNEL
            )
            channel.setMethodCallHandler(AzureLivenessPlugin(activity))
        }
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "startLivenessSession" -> {
                val authToken = call.argument<String>("authToken")
                if (authToken == null) {
                    result.error("INVALID_ARGS", "authToken is required", null)
                    return
                }
                startLivenessSession(authToken, result)
            }
            else -> result.notImplemented()
        }
    }

    private fun startLivenessSession(authToken: String, result: MethodChannel.Result) {
        // Native Azure Face Liveness SDK integration:
        //
        // 1. Initialize FaceSessionClient:
        //    val sessionClient = FaceSessionClientBuilder()
        //        .endpoint("{AZURE_ENDPOINT}")
        //        .buildFaceSessionClient()
        //
        // 2. Create liveness with verify session:
        //    val parameters = CreateLivenessWithVerifySessionParameters()
        //    parameters.authToken = authToken
        //    parameters.deviceCorrelationId = UUID.randomUUID().toString()
        //    sessionClient.createLivenessWithVerifySession(parameters)
        //
        // 3. Launch native camera UI:
        //    val intent = LivenessWithVerifySessionIntents
        //        .createSessionIntent(activity, authToken)
        //    activity.startActivityForResult(intent, LIVENESS_REQUEST_CODE)
        //
        // 4. Handle result in onActivityResult:
        //    val outcome = LivenessWithVerifySessionIntents
        //        .getLivenessWithVerifySessionOperationResultFromIntent(data)
        //    val verifySessionId = outcome.sessionId
        //    val payload = outcome.encryptedVerificationPayload
        //
        // This stub returns a simulated success for now.
        result.success(authToken)
    }
}
